### ISSUE TITLE
Creating a port of Monocypher to Rust

### ISSUE THREAD

**Issue by ethindp** [LoupVaillant/Monocypher#153]
Hello there,
I am currently working on an OS (just a hobby of mine). Sooner or later I'd like to implement a cryptography layer in my kernel, and I love Monocypher; its API is clean, its (not exactly) easy to misuse (from what I know), its code is auditable and its small (and trivial to embed). The issue is that I am no cryptographer. I know a lot about cryptography, especially about how to use algorithms safely, but I am not a cryptographic engineer. Monocypher doesn't look overly difficult to port; I could probably (in my port) simplify theAPI a little. The issue, of course, is what type to use (for, say, Argon2I). By "simplifying the API" I mean something like this. These C declarations:
```C
void crypto_lock(uint8_t        mac[16],
                 uint8_t       *cipher_text,
                 const uint8_t  key[32],
                 const uint8_t  nonce[24],
                 const uint8_t *plain_text, size_t text_size);
int crypto_unlock(uint8_t       *plain_text,
                  const uint8_t  key[32],
                  const uint8_t  nonce[24],
                  const uint8_t  mac[16],
                  const uint8_t *cipher_text, size_t text_size);
```
becomes something like this (in Rust):
```rust
fn lock(mac: [u8; 16], key: [u8; 32], nonce: [u8; 24], plain_text: Vec<u8>) -> Result<Vec<u8>, CryptoError>;
fn unlock(key: [u8; 32], nonce[u8; 24], mac: [u8; 16], cipher_text: Vec<u8>) -> Result<Vec<u8>, CryptoError>;
```
The issue is I don't know how safe this would be. I can implement, for example, the [Drop](https://doc.rust-lang.org/std/ops/trait.Drop.html) trait to make the system automatically wipe cryptographic data (keys, ...) after they go out of scope, potentially making the crypto_wipe function entirely unnecessary. I have a few questions though:

1. How would I go about doing this safely? As I said, I'm no expert in the field of cryptography, and don't really know what's "safe" and what isn't "safe" when implementing cryptographic libs. I've never written an AES cipher either.
2. Who should maintain this, if I do do it? I am in college right now, and I would feel better if it was maintained and audited by @LoupVaillant (who maintains Monocypher as well) or someone else who is more skilled than I am. My kernel, once I got this implemented, would also need to remain in sync, which would be easier if this were an external crate.

I'd also like to here other thoughts about this and how we might go about this. I'm definitely open to suggestions.

Edit: I am aware that there are crates like Ring in Rust which are currently in development. However, using something like Monocypher seems like a saner idea, given that its quite stable and known to be secure already.

**Comment by LoupVaillant** [comment]
> These C declarations: […] becomes something like this (in Rust): […]

Hmm, those function types have two problems:

- `lock()` cannot fail, you don't need to return `Result`. Returning a `Vec<u8>` will simplify user code.
- If you want to prevent misuse, you want your key to have its own type. More generally, Monocypher has several types of keys:

   - Random seed (not really a key)
   - Symmetric (encryption & authentication)
   - Signing (private)
   - Signing (public)
   - Key exchange (private)
   - Key exchange (public)
   - Shared secret (key exchange)

  The keys you need must either be obtained through the proper cryptographic primitive, or converted explicitly. For instance, a shared secret must be hashed before you get a symmetric key. Doing it this way should prevent most misuses.

Wipes are less important in Rust than they are in C. The primary way to exploit unwiped data is to take control of the program, and *then* read all its memory. C and C++ programs have many such bugs. Rust programs are only vulnerable in `unsafe` sections (and external C/C++ libraries). That said, wiping is still a good idea. `Drop` is an excellent choice, especially combined with the idea of giving types to keys, but you probably need `crypto_wipe()` (or a Rust equivalent, if it exists) under the hood nevertheless: if the only thing your destructor does is zero out memory, the compiler is likely to notice this memory is never used again, and optimise the write away, nullifying your efforts.

> How would I go about doing this safely?

Assuming we're talking about a re-implementation, and not a simple binding (one [already exists](https://crates.io/crates/monocypher)), the following should be pretty safe:

- Copy the Monocypher test suite, including all test vectors, and make sure your library passes everything.
- Transliterate bignum arithmetic (Poly1305, elliptic curves) without being clever: this stuff is hard to get right, and limb overflow are very hard to test. (Personally, I stole elliptic curve arithmetic from SUPERCOP Ref10).
- Take advantage of Rust's types. Proper types have the potential to prevent lots and lots of misuses, as well as accidental argument swappings that would be tedious to debug. I didn't do it in C because it isn't the way, but in languages like Rust, ML, Haskell… you can easily have strong type safety.

> Who should maintain this, if I do do it?

Hmm, that's not easy. I could do it, but right now there's quite a bit of work pending in Monocypher. Yes, Monocypher is pretty stable, but we are still missing a couple important functions (Elligator2 support, PAKE, and maybe a safe incremental API for AEAD (the unsafe one just got removed)). I won't be able to duplicate that work on my own. I should be able however to maintain a fork for bugfixes only, once the main repository have achieved stability.


> [Monocypher is] quite stable and known to be secure already.

Thanks for the compliment, but "known to be secure" is maybe a stretch? Was the safety of Monocypher reported to you by a third party, or did you judged the project yourself? (Either is good, I'm just curious.)

Just so you know, I'm currently trying to have Monocypher audited by a security company. Once that's done, Monocypher's security will be much easier to communicate.

**Comment by ethindp** [comment]
For the types, what would those structures look like for custom key types? I could do a SemetricKey, PrivateKey, PublicKey, SharedSecret, PrivateExchange, PublicExchange, and RandomSeed struct, but I'll need to know what they should contain.
Zeroing out memory isn't that difficult to do. If I acquire a pointer to the sensitive data (i.e. key material) and I already know the length, I can just use offset() and write_volatile() to force the compiler to not eliminate the zeros. There's probably a safe way to do this as well.
As for the bignum stuff, how "big" are we talking in regards to bits? Rust has 128-bit integers, both signed and unsigned. Going any larger will require either multiple 128-bit integers or something else like architecture-based intrinsics. And, of course, I'd make sure the tests passed.
As for the "safety" of monocypher, if memory serves the website at one time or another had an audit report. I may be thinking of another library. But still, as I said, I like the library because its not overcomplicated or bloated (OpenSSL) and has a small attack surface. Perhaps I'm wrong on that last point.

**Comment by LoupVaillant** [comment]
> For the types, what would those structures look like for custom key types?

Just an [u8; 32], I guess? The point isn't what the key contains, the point is to make the key type opaque so users only use it for its intended purposes. For instance, you probably don't want users to call `lock()` with a raw shared secret as key. You want to hash that key first. Hiding the 32 bytes of shared secret behind a type helps you warn users about that mistake.

That said, I'd need some time to work out exactly which operations should be permitted, which should be forbidden, and how types can help us enforce the relevant invariants.

> Zeroing out memory isn't that difficult to do

Cool.

> As for the bignum stuff, how "big" are we talking in regards to bits?

Poly1305 multiplies 130 bit  integers together. Curve25519 multiplies 255 bit integers together. Neither fit in any machine word, so we chop up those numbers in "limbs", and perform schoolbook multiplication on them. I have a [tutorial](http://loup-vaillant.fr/tutorials/poly1305-design) on the subject (scroll down to "Cheating at modular arithmetic"). To make things worse, most implementations (including Monocypher) delay the carry operations, because it's much faster. The delayed carry is the most dangerous part: wrong code can fail to carry as much as it needs to, which can trigger an internal "limb overflow" which occurs very rarely, in a way that random tests wouldn't easily catch. When you write that type of code on your own, your only realistic choice is to [prove it works](https://monocypher.org/quality-assurance/poly1305-proof).

> As for the "safety" of monocypher, if memory serves the website at one time or another had an audit report.

Hmm, must be another library. You're right about the attack surface, though. I keep Monocypher small (under 2K lines of code) to minimise errors, so it does have a small attack surface. At least, much smaller than Libsodium or (goodness forbid) OpenSSL.

**Comment by LoupVaillant** [comment]
I'm guessing I answered the question? I'll close this for now, don't hesitate to (ask me to) reopen it if you need more.

**Comment by ethindp** [comment]
Bringing up this issue again because I may have a way of solving the type problems. Would using architecture intrinsics solve the chop issue? It would make the library (temporarily) x86-specific, but since I'll be implementing it in my kernel -- which is currently x86-only -- that wouldn't matter anyway. The intrinsics I would be using would be in the [core::arch](https://doc.rust-lang.org/core/arch/index.html) module. Thoughts on that?

**Comment by LoupVaillant** [comment]
That's one way to do it, but you'd still have to multiply two 130-bit integers together, so you theoretically need up to 259 bits… You can use 128-bit integers to good effect, but then you need two main 64-bit limbs instead of 4 32-bit limbs (and 2 more bits of course). I recall having tried it though, it wasn't really faster.

The fastest method I know of though is described [on my blog](http://loup-vaillant.fr/tutorials/poly1305-design) (under the "Parallelism" section). With SIMD intrinsics, you should be able to process 4 blocks in parallel, and be roughly twice as fast as Monocypher. Can't use 32-bit limbs with this method though (can't benefit from trimmed r any more), so you need 5 26-bit limbs instead.

That said, I'd advise you to optimise Chacha20 first, which is both simpler to do, and the real bottleneck here. (On my machine, Poly1305 is over twice as fast as Chacha20, and SIMD multiplies its speed by about 5 —see the [benchmarks](https://monocypher.org/speed).)
